[RFC] GSoC2015 Improve GMM API

xuewei4d opened the pull request:
GSoC 2015 project,  Improve GMM API
1. [x] Completes [derivation report](https://www.dropbox.com/s/8hlbb7dlwllwcry/VBGMM.pdf?dl=0).
2. [x] Adds new classes. One abstract class `_BasesMixture`. Three derived classes `GaussianMixture`, `BayesianGaussianMixture`, `DirichletProcessGaussianMixture`
3. [x] Decouples large functions, especially in `DirichletProcessGaussianMixture` and `BayesianGaussianMixture`
4. [x] Removes numerical stability fixes for HMM. It seems that whenever there is a numerical issue, the code always adds 10_EPS in the computation. I think in some cases there is a better way to address the problem, such as normalization the extremely small variables earlier, or we just simply remove ```10_EPS``` which is only for HMM.
5. [x] Writes updating functions for `BayesianGaussianMixtureModel` and `DirichletProcessGaussianMixtureModel` according to the report.
6. [x] Provides methods that allow users to initialize the model with user-provided data
7. [x] Corrects kmeans initialization. It is weird when using kmeans initialization, only means is initialized. The weights and covariances are initialized by averaging.
8. [x] Writes several checking functions for the initialization data
9. [x] Adjusts the verbose messages. When `verbose>1`, it display log-likelihood and time used in the same line of the message `Iteration x`
10. [x] Remove `fit_predict`
11. [x] Adds warning for `params!='wmc'`
12. [ ] Studies and contrasts the convergence of classical MLE / EM GMM with Bayesian GMM against the number of samples and the number of components
13. [x] Friendly warning and error messages, or automatically addressing if possible (e.g. random re-init of singular components)
14. [ ] Examples that shows how models can over-fit by comparing likelihood on training and validation sets (normalized by
    the number of samples). For instance extend the BIC score example with a cross-validated likelihood plot
15. [x] Testing on 1-D dimensions
16. [x] Testing on Degenerating cases
17. [optional] add a partial_fit function for incremental / out-of-core fitting of (classical) GMM, for instance http://arxiv.org/abs/0712.4273 
18. [optional] ledoit_wolf covariance estimation or `PCA(n_components).fit(X_train).get_covariance()`

xuewei4d commented:
I have not make ```KDE``` to be based on ```DensityMixin```. @ogrisel @lesteve @amueller .

amueller commented on sklearn/mixture/gmm.py:
```diff
@@ -329,23 +326,42 @@ def score_samples(self, X):
+        logprob, _ = self._lp_resp_samples(X)
         return logprob
 
+    def responsibility_samples(self, X):
```
what is this method for? Why not just use ``predict_proba``? We would rather not add to many functions to the API.

xuewei4d replied:
Sorry. I didn't realize that we have ```predict_proba```.

amueller commented:
The thing with changing ``score_samples`` is that it is an incompatible change. So we have to a) accept that we break user code (rather not) or b) rename the estimator to ``GaussianMixtureModel`` or something, and make the old class behave like it did, deprecate it, and make the new class behave as we want.

xuewei4d commented:
Sorry for my bone-headness. I will build a new Class ```GaussianMixtureModel```. And we should add deprecate messages in ```score_samples``` right?

amueller commented:
Rename the old class to the new name, then implement GMM by inheriting from the new class and overwriting score_samples. Then deprecate the whole "old" GMM. At least I think that was the consensus.

xuewei4d commented:
Did you mean to rename ```class GMM``` to ```class GaussianMixtureModel``` and implement ```GMM``` in the same file, i.e. gmm.py?

amueller commented:
yes, that sounds good.

vene commented on sklearn/mixture/tests/test_gmm.py:
```diff
@@ -471,3 +470,33 @@ def test_verbose_second_level():
+    assert_warns_message(DeprecationWarning,
+                         "The 'score_samples' method in GMM has been "
+                         "deprecated in 0.17 and will be removed in 0.19."
+                         "Use 'score_samples' method "
```
*the* 'score_samples' method, *the* 'log_probability', *the* 'predict_proba' method, *the* 'responsibility'. Same in all deprecation warnings.

Also I think '*of* each sample' sounds better than '*for* each sample' in this context.

vene commented on sklearn/mixture/gmm.py:
```diff
@@ -819,3 +816,230 @@ def _covar_mstep_tied(gmm, X, responsibilities, weighted_X_sum, norm,
+                      "will be removed in 0.19. "
+                      "Use the GaussianMixtureModel instead.",
+                      DeprecationWarning)
+        if params is not 'wmc':
```
Is this supposed to test for the default value?

In this case, the default value should be set to `None`, this should be `if params is not None:`, and then set it to `'wmc'` if `None`.

Though I don't think it's safe to do that in `__init__`, it should probably be in `fit`.

Or we could leave it this way, but if user code explicitly says `GaussianMixtureModel(params='wmc')`, it won't raise a warning. I don't like this: I often spell out parameters even if they're at the default value, just to make my code more readable. It would lead to code later breaking without ever having raised a warning.

xuewei4d replied:
Thanks for your comments. Neither in ```__init__``` nor in ```_fit```  setting to ```'wmc'```could  pass the tests. So I have to do that in the ```m_step```.

xuewei4d committed:
Adds DensityMixin

Adds the new classes

Updates new classes

Fixes bugs and typos

1. completes check functions
2. completes initialization functions
3. corrects the time to compute log-likelihood
4. makes verbose messages and its displaying more clear
5. renames modules
6. removes hmm numerical stability fixes '+ 10*EPS'
7. simplifies 'fit_predict'
8. PEP8

1. Rename
2. Ajusts the verbose messages

1. Fixes bugs of checking methods
2. Completes methods for 'diag' cases

1. Fixes RNG issue
2. Addes the method name to initialize to the verbose message
3. Completes the updating functions for tied, spherical. Slightly optimized.

1. Improves the flow of ```fit```
2. Doc and test
3. Fixes bugs

1. Fixes bugs
2. Adds test cases

1. Addes more tests
2. Fixes bugs
3. Docs

1. Fixes bugs
2. Adds more tests
3. Docs

Minor fix

1. Refactor ```MixtureBase```
2. Adds various methods into ```BayesianGaussianMixture```
3. Removes ```fit_predict```
4. Fixes ```DensityMixin```

1. Fixes bugs
2. Complete BayesianGaussianMixture for 'full' case

1. Fixes bugs
2. Completes the computation of lower bound

Split gaussianmixture.py

Refactoring

1. snapshot method for debug
2. DirichletProcessGaussianMixture

BayesianGaussianMixture for tied covariance/precision

1 Bug fixes
2 ```BayesianGaussianMixture``` for diag covariance

lower bound for dpgmm

BayesianGaussianMixture for spherical covariance

AIC, BIC, score for GaussianMixture
score for BayesianGaussianMixture

fixes typoes
rename parameters

cosmit
simplify _estimate_suffstat

COSMIT

bug fixes and test cases

bug fixes and test cases

bug fixed and test cases

test cases

test cases

test cases

test cases for dpgmm and rename prior parameters

verbose msg

PEP8 and PEP257

fixes score, replace lowerbound with log probabilities in BGM

remove comments

doc and fix bugs

fix some numerical issue

doc

xuewei4d commented:
I am sorry I didn't get a very clear mind in the beginning, and there is a lot of work on refactoring ```GMM```. So I removed the ```GaussianMixtureModel``` I created for the inconsistency problem about ```score```, but I created a new class in an separated file, just an prototype. Here is that my thought.

* added the parameters ```weights, means, covars``` in ```_init_``` to allow users initialize the model
* added methods to check user-provided initialization data
* added methods that compute the sufficient statistics of the data, which is used in m-step. They are N_k, N_k * \bar{x}_k, N_k * S_k.
* added estimation functions as abstract methods to the base class ```_MixtureModelBase``` such that ```GaussianMixtureModel```, ```BayesianMixtureModel```, ```DirichletProcessMixtureModel``` could implement their own updating functions.
* added ```DensityMixin``` to compute ```score``` for all density estimators

The code is in the earliest version. Any feedback is welcome.

vene commented:
I may be a bit late to the party but if these classes get rename, may I object to having "Model" in the name, and suggest simply `GaussianMixture` etc? I mean, most of our estimators are models of some sort, the word doesn't bring much, and the docstrings will still say "Gaussian Mixture Model" and "GMM" for searchability.

xuewei4d commented:
@vene from my perspective, the name 'GaussianMixtureModel' is OK, because in the literature everybody uses this name. The word 'model' is kind of a part of it.

xuewei4d commented:
I am confused that recent PR #4579. How can zero-filled responsibility is a kind of initialization? [Line571](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/mixture/gmm.py#L571). The L1-norm of each responsibility must be 1. If we does need quick-initialization, we should initialize the model parameters, weights_, means_, covariances_, not responsibility.

>  # self.n_iter == 0 occurs when using GMM within HMM
            # Need to make sure that there are responsibilities to output
            # Output zeros because it was just a quick initialization

I cannot find somewhere needs responsibilities, except ```self.fit_predict``` merged from the PR #4579. 
HMM of sklearn is gone. hmmlearn does not set ```n_iter=0``` (https://github.com/hmmlearn/hmmlearn/blob/master/hmmlearn/hmm.py#L592) and ```_fit``` is a private method, which should not be called by other modules. So returning responsibility is only for ```self.fit_predict```. However, this method directly use the responsibility of ```_fit```, which will skip the last EM iteration, because the returned responsibility is computed with parameters optimized by the last two iteration. 

And [Line442](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/mixture/gmm.py#L442) has a weird parameter ```do_prediction=False```??

clorenz7 commented:
Yes, do_prediction is not needed, I apologize for my oversight.

There was discussion in the PR pre-merge about the removal of HMM, and the
not-needed n_iter=0. It is not that another module would be calling
self._fit, but that a module which subclassed from it would make use of it.
The addition was largely defensive since responsibilities variable was set
in the for loop over n_iter. The responsibilities being all zeros as a
degenerate case seemed ok since a GMM fit with n_iter=0 is already
degenerate.

You are right about the responsibility returning only being for
fit_predict, and it skipping the last EM iteration. I wrote a comment about
that:
```
Warning: due to the final maximization step in the EM algorithm, with low
iterations the prediction may not be 100% accurate ```
During the development of the method, the feedback that I received was that
that the idea of the fit_predict method is to save time over someone just
calling ```model.fit``` then ```model.predict```.







On Wed, Jun 17, 2015 at 11:10 AM, Wei Xue <notifications@github.com> wrote:

> I am confused with the recent PR #4579
> <https://github.com/scikit-learn/scikit-learn/issues/4579> @clorenz7
> <https://github.com/clorenz7> . How can zero-filled responsibility is a
> kind of initialization? Line571
> <https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/mixture/gmm.py#L571>.
> The L1-norm of each responsibility must be 1. If we does need
> quick-initialization, we should initialize the model parameters, weights_,
> means_, covariances_, not responsibility.
>
> # self.n_iter == 0 occurs when using GMM within HMM
> # Need to make sure that there are responsibilities to output
> # Output zeros because it was just a quick initialization
>
> I cannot find somewhere needs responsibilities, except self.fit_predict
> merged from the PR #4579
> <https://github.com/scikit-learn/scikit-learn/issues/4579>.
> HMM of sklearn is gone. hmmlearn does not set n_iter=0
> <https://github.com/hmmlearn/hmmlearn/blob/master/hmmlearn/hmm.py#L592>.
> _fit is a private method, which should not be called by other modules. So
> returning responsibility is only for self.fit_predict. However, this
> method directly use the responsibility of _fit, which will skip the last
> EM iteration, because the returned responsibility is computed with
> parameters optimized by the last two iteration.
>
> And Line442
> <https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/mixture/gmm.py#L442>
> has a weird parameter do_prediction=False??
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/4802#issuecomment-112896427>
> .
>

ogrisel commented:
+1 for simplifying the code by taking the removal of HMM into account and not dealing with the `n_iter=0` case that should never appear.

ogrisel commented:
I am fine with `GaussianMixture`, `BayesianGaussianMixture` and `DirichletProcessGaussianMixture` if we want shorter class names.

xuewei4d commented:
OK. @ogrisel . I have renamed them. Now I rearrange way of displaying the verbose messages and computing the log-likelihood. It now goes like as follows, when ```verbose=1```
```
The estimation of GaussianMixture is started.
Initialization 1 
	Iteration 1 
	Iteration 2 
	Iteration 3 
	Iteration 4 
	Iteration 5 
	Iteration 6 
	Estimation converged.
Initialization 2 
	Iteration 1 
	Iteration 2 
	Iteration 3 
	Iteration 4 
	Iteration 5 
	Iteration 6 
	Estimation converged.
```

Verbose=2

```
The estimation of GaussianMixture is started.
Initialization 1	Initializing parameters ... 
	Weights are initialized.
	Means are initialized.
	Covariances are initialized.
	Used 0.03758s	Log-likelihood -53339.5745805 
	Iteration 1 	used 0.04458s 	Log-likelihood -53338.6651993 
	Iteration 2 	used 0.05161s 	Log-likelihood -53338.6551914 
	Iteration 3 	used 0.06380s 	Log-likelihood -53338.6550089 
	Iteration 4 	used 0.07336s 	Log-likelihood -53338.6550055 
	Initialization 1 used 0.07918s 	Estimation converged.
	Better parameters are found.
Initialization 2	Initializing parameters ... 
	Weights are initialized.
	Means are initialized.
	Covariances are initialized.
	Used 0.03274s	Log-likelihood -53339.5745805 
	Iteration 1 	used 0.04651s 	Log-likelihood -53338.6651993 
	Iteration 2 	used 0.05707s 	Log-likelihood -53338.6551914 
	Iteration 3 	used 0.06348s 	Log-likelihood -53338.6550089 
	Iteration 4 	used 0.07265s 	Log-likelihood -53338.6550055 
	Initialization 2 used 0.07833s 	Estimation converged.
```

If it is not converged, raises warnings

```
The estimation of GaussianMixture is started.
Initialization 1
	Iteration 1 
	Iteration 2 
Initialization 2
sklearn/mixture/gaussianmixture.py:430: UserWarning: Iteration is not converged.Try to increase the number of iterations
  if not self.converged_:
sklearn/mixture/gaussianmixture.py:430: UserWarning: Iteration is not converged.Try to increase the number of iterations
  if not self.converged_:
	Iteration 1 
	Iteration 2 
```

verbose=2
```
The estimation of GaussianMixture is started.
Initialization 1	Initializing parameters ... 
	Weights are initialized.
	Means are initialized.
	Covariances are initialized.
	Used 0.03542s	Log-likelihood -53083.5098991 
	Iteration 1 	used 0.04219s 	Log-likelihood -53081.1306632 
	Iteration 2 	used 0.04858s 	Log-likelihood -53081.043168
	Initialization 1 used 0.05178s 
	Better parameters are found.
Initialization 2	Initializing parameters ... 
	Weights are initialized.
	Means are initialized.
	Covariances are initialized.
	Used 0.03971s	Log-likelihood -53083.5098991 
	Iteration 1 	used 0.04711s 	Log-likelihood -53081.1306632
sklearn/mixture/gaussianmixture.py:430: UserWarning: Iteration is not converged.Try to increase the number of iterations
  if not self.converged_:
sklearn/mixture/gaussianmixture.py:430: UserWarning: Iteration is not converged.Try to increase the number of iterations
  if not self.converged_:
 
	Iteration 2 	used 0.05413s 	Log-likelihood -53081.043168
	Initialization 2 used 0.05824s 

```

xuewei4d commented:
@amueller @ogrisel @lesteve @vene 
I am not sure about whether we should deprecate ```params``` which could make algorithm to estimate a part of parameters #4429, #3945. For example,  if ```params='wm' ```, then the algorithm would skip the estimation of covariances. 

Indeed, when ```params='wm' ```, the algorithm cannot run correctly. But what if the user provides the ground truth to the model and sufficient data representing the underlying data distribution, so he just would like to estimate the other parameters e.g. 'wm' ? In this case, we may have to reserve ```params```, and raise a warning that let users know setting ```params='wm'``` is on their own risk.

ogrisel commented on sklearn/mixture/gaussianmixture.py:
```diff
@@ -0,0 +1,685 @@
+            if not self.converged_:
+                warnings.warn('Initialization %s is not converged. '
+                              'Try to increase the number of iterations'
+                              % (init + 1))
```
Instead of using the default `UserWarning` class you should use the dedicated `ConvergenceWarning` class of scikit-learn here.

ogrisel replied:
Also it's better to use `%d` to explicitly format integer values.

xuewei4d replied:
Okay.

ogrisel commented on sklearn/mixture/gaussianmixture.py:
```diff
@@ -0,0 +1,685 @@
+                                                  self.n_components)
+                responsibilities = (responsibilities /
+                                    responsibilities.sum(axis=1)
+                                    [:, np.newaxis])
```
I would also put a log here for the initialization of the components, either via uniform random assignments of the samples or via kmeans.

xuewei4d replied:
Do you mean the log value of the responsibility?

ogrisel commented on sklearn/mixture/gaussianmixture.py:
```diff
@@ -0,0 +1,685 @@
+                # other initialization methods as long as
+                # they return responsibilities
+                responsibilities = np.random.rand(X.shape[0],
+                                                  self.n_components)
```
You should pass a `random_state` instance of the `np.random.RandomState` class to the `_initialize` method instead of using the PRNG singleton of the `np.random` module. 

The `random_state` local variable should be initialized at the beginning of the `fit` method from the `self.random_state` constructor parameter using the following idiom:

```python
random_state = check_random_state(self.random_state)
```

See: http://scikit-learn.org/dev/developers/contributing.html#random-numbers

The goal is to make it possible for the user to get deterministic outcomes by fixing the PRNG seed while preserving the fact that the `fit` method should be idempotent (in particular the constructor parameters should not be mutated by a call to `fit`).

xuewei4d replied:
Okay.

ogrisel commented:
About the bugs when `params!='wmc'` I am not sure. The git history does not show who originally contributed that feature because it predates the svn to git migration.

I wonder it partial estimation of the mixture parameters is really a use case. If it is how could we make sure that we are not making any mistakes if only some parameters are to be estimated while others are held to some fixed values. Maybe @bthirion has an opinion on those issues?

bthirion commented:
In principle, not re-estimating covariance or weights makes the algorithm robust to numerical singularities when there are few data points. Hence, I think that this is useful. 
I'm less sure about the means (following #4429).  The use case is indeed to convert a clustering model to a probabilistic one at little or no cost. I don't know whether this is outside the scope of skl.

 Regarding covariance estimate with fixed means: If I'm not mistaken, you simply need to update the covariance estimation with a term that takes into account the difference between samples mean and fixed mean.

Here is a paper that I consider as a reference for GMMs: https://hal.archives-ouvertes.fr/hal-00469477/document
Note that their model is very complete, more than necessary. At least it gives you ides of all the relevant parametrizations. HTH

lesteve commented on sklearn/mixture/gaussianmixture.py:
```diff
@@ -0,0 +1,685 @@
+                print_('\n\tInitialization %s used %.5fs' %
+                       (init + 1, time() - start_init_time), end='')
+            if self.verbose > 0:
+                print
```
Is this print really necessary ? In case it is, I guess it should be replaced by print_() or something. At the moment, the behaviour is going to be different between python 2 (print a new line) and 3 (do nothing).

xuewei4d replied:
Yes. It should be ```print_```. Maybe I will rewrite message printing part. It somewhat messed the ```fit``` method.

xuewei4d commented:
@ogrisel I think the code about 'params' is from the very first commit by @fabianp.  The commit message says
> Just made some trivial changes to import under python2.5
Original can be found here: http://github.com/ronw/gm From: Ron Weiss <ronweiss@gmail.com>

I can explain the bugs in the covariance estimation. In the method ```_covar_mstep_full``` and ```_covar_mstep_tied```.  The covariance should be estimated by 
![cov_true](https://cloud.githubusercontent.com/assets/1180956/8285801/c29d0086-18d3-11e5-9083-2936f1acb941.png)
Assuming 
![covar_simple](https://cloud.githubusercontent.com/assets/1180956/8285811/d3de2abe-18d3-11e5-91fa-05955ce86bad.png)
which is not true when ```params!='wmc'```.
Then the estimation equation becomes
![covar_false](https://cloud.githubusercontent.com/assets/1180956/8285836/02fa666e-18d4-11e5-8b43-9f9575b92c91.png)
which are used in ```_covar_mstep_full``` and ```_covar_mstep_tied```.

The use cases  that users providing ground truth sounds like very rare cases. I didn't see any other implementation neither in mclust nor in MIXMOD. I slightly prefer to deprecate that now.

amueller commented:
I feel that ``params`` is outside the scope of sklearn and it complicates the code too much. In particular if we want to also have it for the bayesian variants, it will be very tricky.

If you want a more specific model, it is really not that hard to implement yourself for your usecase.
I think we should allow passing parameters for init, say ``init_means=some_array``, but then keep updating them.

bthirion commented:
This sounds like a pragmatic solution. +1

xuewei4d commented:
Yeah. I have already done that in ```__init__``` that allows users to pass weights, means, covariances

xuewei4d commented:
Hi, I just fixed a bug that computes log-likelihood twice. ```_estimate_responsibilities``` is changed to ```_estimate_log_probabilities_responsibilities````. 

For the computing log-probabilties for 'tied' and 'spherical' case, I fixed the bug reported in #4109, #4039, #4036. Now we don't have to call the function for 'full' case to get the log probabilities for 'tied' case.

xuewei4d commented:
I am wondering why ```n_init``` is called number of initialization? Should ```n_run``` be better ?

amueller commented:
well it is the number of random initializations tried, right?

xuewei4d commented:
I don't think so. The fit function is a nested for loop. The outer loop iterates over the number of trials. The inner loop iterates over the EM-like optimization. The initialization runs only once in the outer loop.

amueller commented:
well depends what you call a trial and what an initialization. It runs the full algorithm with ``n_init`` different initializations, right? This is the same meaning as in ``KMeans`` (but not the same as in MiniBatchKMeans I think).

amueller commented:
I think this is most commonly known as "random restarts". I haven't heard the word ``run`` in this context.

xuewei4d commented:
@amueller OK. I will keep ```n_init```.

xuewei4d commented:
I have a question about testing. There are some global checking functions. So I write a unit test for each function, in which the functions are called directly and check if it raises the correct exceptions. But do I need to do functional test? In other words, I create ```gaussianmixture``` object and let some functions of it to call these checking functions, and check the exceptions.

xuewei4d commented:
Hi, I don't understand which is the test for [L247](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/mixture/tests/test_gmm.py#L247). Could you explain it @ogrisel ? It is contributed by @ronw at the first.

ogrisel commented on sklearn/base.py:
```diff
@@ -449,6 +449,25 @@ def fit_transform(self, X, y=None, **fit_params):
+    """Mixin class for all density estimators in scikit-learn."""
+    _estimator_type = "DensityEstimator"
+
+    def score(self, X, y = None):
```
pep8 style: no whitespaces around `=` to when used for keyword arguments.

ogrisel commented on sklearn/base.py:
```diff
@@ -449,6 +449,25 @@ def fit_transform(self, X, y=None, **fit_params):
+        score: float
+            sum of log probability of samples
+        """
+        return np.sum(self.score_samples(X))
```
I wonder if we should not use the "mean" log likelihood as the `score` method for the density models. Otherwise we will have a `score` function that is not [statistically consistent](https://en.wikipedia.org/wiki/Consistency_%28statistics%29): it will grow unbounded as you add more test samples which is a very unexpected behavior and is not consistent with the default score method for classifier (mean accuracy) and regressors (empirical r2).

In particular if the `score` function is used to evaluate overfitting, the training score and the test score would have to be evaluated on a training set and a test set with the same number of samples, otherwise the results cannot be compared.

My proposal would be to do use instead: `np.mean(self.score_samples(X))`. If we use that as the score we should make it clear that this is the mean log-likelihood instead of the total log likelihood.

What do people (@xuewei4d, @amueller, @jakevdp) think of that suggestion?

xuewei4d replied:
I agree. This ```score``` could be different from the log-likelihood of the model on the training data. If we use ```mean```, we should let users know that, and provide another method to return the log-likelihood.

ogrisel replied:
`score_samples` already does that (the user can always call sum by him/her self to get the likelihood of a batch).

I find `score_samples` not a very good name, maybe we should call rename the `score_samples` method `log_likelihood`.

xuewei4d replied:
likelihood is often referred as a function of parameters. For individual data samples, I think we should call the same numerical value as log probabilities.

ogrisel commented on sklearn/mixture/gaussianmixture.py:
```diff
@@ -0,0 +1,1065 @@
+    # check range
+    if (any(np.less(weights, 0)) or
+            any(np.greater(weights, 1))):
+        raise ValueError("The 'weights' should be in the range [0, 1]")
```
Please be informative and report observed min and max weights.

ogrisel commented on sklearn/mixture/gaussianmixture.py:
```diff
@@ -0,0 +1,1065 @@
+
+    # check normalization
+    if not np.allclose(np.abs(1 - np.sum(weights)), 0.0):
+        raise ValueError("The 'weights' should be normalized")
```
Please report the sum of the weights in the error message.

ogrisel commented on sklearn/mixture/tests/test_gaussianmixture.py:
```diff
@@ -0,0 +1,557 @@
+                         covariance_type=covariance_type)
+
+
+def test__check_X():
```
why the double `_`? Better be consistent with the project naming conventions. We don't use double `_` in other test function names.

xuewei4d replied:
That's because we are testing ```_check_X```.

ogrisel replied:
Ah ok, I would still collapse the two consecutive `_` but that's fine.

xuewei4d commented:
I am going to code ```BayesianGaussianMixture```. The m-step updating functions have something in common with ```GaussianMixture```. I may have to decompose ```_estimate_log_probabilities_full``` and else into ```mahala_dist``` and ```log_det``` or something.

ogrisel commented:
> I am going to code `BayesianGaussianMixture`. The m-step updating functions have something in common with `GaussianMixture`. I may have to decompose `_estimate_log_probabilities_full` and else into `mahala_dist` and `log_det` or something.

I don't understand this sentence. Could you please rephrase and give more details on this planned change?

> I also have to remove the fast covariance estimation for 'tied', because we still need old GMM that allow users to set params='wm', and it cannot be reused in BayesianGaussianMixture

I think you can move on by dropping the support for the `params` hyperparam for now (all the parameters are estimated). We can later re-think what would need to be changed to re-add that feature correctly when it makes sense only.

ogrisel commented:
> Hi, I don't understand which is the test for  [L247](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/mixture/tests/test_gmm.py#L247). Could you explain it @ogrisel ? It is contributed by @ronw at the first.

The comment says it all: the EM algorithm theoretically guarantees an increase in the likelihood of the training set at each iteration. However we don't really implement the real MLE update for the covariance because of the hard-coded `min_covar` offset on the diagonal of the covariance matrices. I don't know where the `-0.5` value comes from. It was probably be found empirically from failing unittest.

This test could be improved by checking that we can use `self.threshold = 0.0` for the same test when we fit a GMM model with `min_covar=0.0` (by choosing a dataset / random_state combination that does not trigger a singularity).

ogrisel commented:
> I have a question about testing. There are some global checking functions. So I write a unit test for each function, in which the functions are called directly and check if it raises the correct exceptions. But do I need to do functional test? In other words, I create gaussianmixture object and let some functions of it to call these checking functions, and check the exceptions.

Yes we need such tests that use the public API of the estimator object. I would rather call them integration tests (especially when several scikit-learn components are tested together). Have a look at other test files in the project to understand what is typically tested in test suites.

Note that we have a battery of standard checks that are applied to all estimators in `sklearn/test/test_common.py` but they don't check that methods actually work (they just check that common expected behavior across all models).

xuewei4d commented:
> The comment says it all: the EM algorithm theoretically guarantees an increase in the likelihood of the training set at each iteration.

I noticed there is a different behavior between this implementation and the current version. After 'gmm' call 'fit', it will use estimated parameters as initialization in the next call of 'fit'. But it won't happen in this PR. The second time call of 'fit' will initialize the parameters like the first call.

xuewei4d commented:
> I don't understand this sentence. Could you please rephrase and give more details on this planned change?

Sorry. I mean the Equation 3.43, 3.48 in my draft are kind of extension to ```log_det``` and ```mahala_dist```, but in ```GaussianMixture``` they are computed in the same method. So I am considering breaking each E-step method into some parts.

ogrisel commented:
> I noticed there is a different behavior between this implementation and the current version. After 'gmm' call 'fit', it will use estimated parameters as initialization in the next call of 'fit'. But it won't happen in this PR. The second time call of 'fit' will initialize the parameters like the first call.

Good catch. The behavior of this PR is the correct behavior. The other behavior could be implemented if the user were allowed to pass a `warm_start=True` hyperparameter. If it's not too complicated to add, I would be +0 to add support for `warm_start=True` as a non-default of option fo GMMs.

ogrisel commented:
> Sorry. I mean the Equation 3.43, 3.48 in my draft are kind of extension to log_det and mahala_dist, but in GaussianMixture they are computed in the same method. So I am considering breaking each E-step method into some parts.

Do as you prefer for now. Once the implementation is working (e.g. tests pass) we can always discuss and refine the code organization.

xuewei4d commented:
Hi, all, I am finishing the updating functions of ```BayesianGaussianMixture```. Since the Equation 7.4 in my draft in the lower bound is almost the same as responsibility. So I am considering to save the responsibility in the class. @ogrisel  @lesteve

ogrisel commented:
> So I am considering to save the responsibility in the class.

You mean when computing the responsibilities of the training samples you want to store that as an attribute on the estimator instance? I think it's a bad idea: we don't want to allocate datastructure that scale with the number of training samples if their are not reused to make prediction on test data.

Just allocate the array of responsibilities as a local variable in the fit method pass those as arguments to the private method of the class when needed. The same array can be modified inplace when doing EM iterations to avoid costly reallocation of that array.

xuewei4d commented:
Yes. I am currently doing that way. After I took a look at [bnpy](https://bitbucket.org/michaelchughes/bnpy-dev/src/98f63d64784f/bnpy/obsmodel/?at=master), I think it is a good idea to decompose the model into the class of observed variables and the class of hidden variables.

xuewei4d commented:
@ogrisel @amueller  @lesteve  I am coding the lower bound but I cannot organize the computing of the terms in lower bound very well.

To compute the lower bound, I need the matrix from Cholesky decomposition of ```inv_W```, ```ln_tilde_Lambda```, ```ln_tilde_w```, log probabilities for each sample, each component, and the responsibility. However, to compute the log probabilities, I also need ```inv_W```, ```ln_tilde_Lambda```, ```ln_tilde_w```; to compute responsibilities I need   log probabilities.  The data that is computed earliest (like ```inv_W``` in the function ```_estimate_log_X_full``` which is called by ```_e_step```) is still needed when computing the latest term (like lower bound).

So it seems that I have to store them somewhere. Of course, saving responsibilities (n_samples x n_features) and log probabilities is not a good idea, but passing them around seems not elegant, needless to say  ```inv_W```, ```ln_tilde_Lambda```

xuewei4d commented:
@ogrisel I am not sure we are on the same page. What I mean is that, for example, in ```_e_step```, I would do ```self.resp_ = self.estimate_resp(X)```.

xuewei4d commented on sklearn/mixture/bayesianmixture.py:
```diff
@@ -0,0 +1,477 @@
+                                  n_features / beta + nu * mahala_dist)
+        log_X -= .5 * (n_features * np.log(2 * np.pi))
+        self.inv_W_chol = inv_W_chol
+        self.log_lambda = log_lambda
```
Hi @ogrisel @amueller @vene @lesteve I didn't save resp and other data which could scale along with n_samples, but I did save other temporary data as instance attributes to compute lower bound when doing e_step. I don't know this is good or bad style.  Any comment?

xuewei4d commented:
OK. I am going to implement a decorator, 'cached_property' to cache the intermediate results when doing e_step.

xuewei4d commented:
I tried to use ```cached_property``` to store the intermediate variables such as, log_pi, log_lambda, but didn't get much benefits. I also tried to split the base class ```MixtureBase``` into three abstract classes ```ObsMixin```, ```HiddenMixin``` and ```MixtureBase(ObsMixn, HiddenMixin)```, and subclass ```ObsGaussianMixin(ObsMixin)```,  ```MixtureMixin(HiddenMixin)```, ```GaussianMixture(MixtureBase, ObsGaussianMixin, MixtureMixin), but Python does allow me to do this because of MRO problem. :-|.

xuewei4d commented:
![demoanimation](https://cloud.githubusercontent.com/assets/1180956/8710149/a6437600-2b15-11e5-9524-d503200f01a5.gif)

A ```BayesianGaussianMixture``` demo on recalled old faith data.

xuewei4d commented:
I implemented ```DirichletProcessGaussianMixture```. But it looks the similar to ```BayesianGaussianMixture``` in terms of behavior.  Both of them can infer the best number of components. ```DirichletProcessGaussianMixture``` took a slightly more iteration than ```BayesianGaussianMixture```, around 60 iterations. 

If we solve Dirichlet Process Mixture by Gibbs sampling, we don't need to specify the 
 truncated level T, which means only ```alpha``` the concentration parameter is enough. In the other hand, with variational inference, we still need to specify the maximal possible number of components, i.e., the truncated level. @ogrisel @lesteve @amueller.

xuewei4d commented:
The lower bound seems a little strange. It is not always going up. When some clusters disappear, it goes down a little bit, then go up straight. I think it is because the estimation of the parameters is ill-posed when these clusters have data samples less than the number of features. Any idea? @ogrisel @lesteve @amueller .
![log_lh](https://cloud.githubusercontent.com/assets/1180956/8734157/125f03c0-2bd6-11e5-9f0a-fe4087d32c97.png)

xuewei4d commented:
The current implementation DPGMM cannot infer the correct parameters, no matter how I set the concentration parameter alpha. @ogrisel @lesteve @amueller .

``` Python
VBGMM(alpha=0.0001, covariance_type='full', init_params='wmc',
   min_covar=None, n_components=6, n_iter=100, params='wmc',
   random_state=None, thresh=None, tol=0.001, verbose=0)
 ```

 It gives only one components. The ```weights_``` is 

 ```
 array([  7.31951611e-07,   7.31951611e-07,   7.31951611e-07,
         7.31951611e-07,   7.31951611e-07,   9.99996340e-01])
 ```

xuewei4d commented:
Hi, I got a problem about the updating function for 'diag' and 'spherical' covariance cases.  The model with these two kinds of covariance use Gamma distribution instead of Wishart distribution.  Gamma distribution has two forms, ![1](https://upload.wikimedia.org/math/9/a/2/9a277651cacd3a06158a9d7800415972.png) and ![2](https://upload.wikimedia.org/math/4/6/2/462d87c22361038f29ada31f31467d7d.png). PRML use the second definition, however this form is somewhat not consistent wit Wishart distribution. ![w](https://upload.wikimedia.org/math/2/5/5/255ae736aa2bdc387ed3134453a4cb9d.png). The parameter \beta in the second form of Gamma distribution is the inverse of the parameter W in Wishart distribution, if you assume W is the Gamma distribution parameter.  Currently, in my PR, the two sets of parameters of Gamma and Wishart  distributions share same names. So to prevent from any confusion, I would use the first definition of Gamma distribution in my derivation draft and the code

xuewei4d commented:
Hi, I just updated my derivation draft to fix the lower bound for DPGMM. I am not so sure about Equation 9.30 and 9.31 in my derivation draft, see the first comment in this PR. Could you help me to verify these equations? @ogrisel @amueller @lesteve

bthirion commented:
Sorry, where is the derivation draft ? Best,

xuewei4d commented:
@bthirion Here it is. https://www.dropbox.com/s/8hlbb7dlwllwcry/VBGMM.pdf?dl=0
Sorry, I thought everybody knows that.

ogrisel commented:
@xuewei4d I think we should leave aside DPGMM for now and focus on finishing GaussianMixture and BayesianGaussianMixture for now. We need:

- make the tests pass and ensure at least 95% coverage of the code
- update the mixture documentation
- update the mixture examples
- benchmark the peformance of GaussianMixture vs the old GMM: we should not have any performance regression (time-wise and space-wise).

For me this is the priority.

xuewei4d commented:
Hi @ogrisel, yes, I am going to do the test cases for BayesianGaussianMixture. I think I finished ```DirichletProcessGaussianMixture```. Now I have implemented all functions of these three models for all covariance types.

xuewei4d commented on sklearn/mixture/bayesianmixture.py:
```diff
@@ -0,0 +1,938 @@
+        return np.sum(.5 * self._log_lambda +
+                      .5 * self.n_features * np.log(self.mu_beta_ / (2 * np.pi))
+                      - .5 * self.n_features
+                      - wishart_entropy)
```
I have a question about PEP8 style. PEP8 said about these lines ```W503 line break before binary operator```. But I think putting the ```-``` operator to the previous line might cause people who would like to check the function against the math equation to omit the minus sign. Placing ```-``` sign in the front of the line would make it noticeable.  @ogrisel

ogrisel replied:
I don't have a strong opinion on this particular point. flake8 seem to ignore it by default on my box.

xuewei4d commented:
Another question is about methods like ```_estimate_log_BGaussian_prob_full``` ```_estimate_p_mu_lambda_full```. For these functions that only about computations, should I make them  as class methods, or as global functions outside of class?

ogrisel commented:
I was reviewing your code and fixing code style issues along the way:

https://github.com/xuewei4d/scikit-learn/pull/2

There has been some concurrent editing so it does not merge directly but by reading the diff you can get the gist of the coding conventions we follow in scikit-learn (PEP8 and PEP257 for the docstrings). You can use a linter such as flake8 to find code style violations and trivial bugs (undefined variables) or other issues (useless imports). If you use and editor such as Atom, install the linter and linter-flake8 packages to have them show up in your editor directly.

ogrisel commented:
More importantly I think this code has too many indirections in as helper methods or help functions with trivial body. Having to hope around the source code too much harms readability. In Python, flat is better than nested. So when helper function are used only once and have a body of a couple of lines it's better to collapse them in the body of the caller. I will try to make more precise recommendations as github inline comments.

ogrisel commented on sklearn/mixture/gaussianmixture.py:
```diff
@@ -0,0 +1,697 @@
+    xk : array-like, shape = (n_components, n_features)
+    """
+    # remove + 10 * EPS
+    return np.dot(resp.T, X) / nk[:, np.newaxis]
```
The body of this function could be collapsed in the `_estimate_suffstat` function to skip a useless indirection when reading the code.

ogrisel commented on sklearn/mixture/gaussianmixture.py:
```diff
@@ -0,0 +1,697 @@
+    -------
+    xk : array-like, shape = (n_components, n_features)
+    """
+    # remove + 10 * EPS
```
is this a TODO? or an old comment that does not apply anymore?

xuewei4d replied:
It's just a reminder for possible numerical issues. I would remove it shortly.

ogrisel commented:
> Another question is about methods like _estimate_log_BGaussian_prob_full _estimate_p_mu_lambda_full. For these functions that only about numerical computations, should I make them as class methods, or as global functions outside of class? 

I think it's fine to use module level (private) functions.

xuewei4d commented:
I fixed almost all of style problems this weekend. @ogrisel

amueller commented:
Sorry for the prolonged silence. I'm just catching up with the issues and I hope to have a look at your PR soon.

amueller commented:
"PEP8 said about these lines W503 line break before binary operator" I disagree with that. @ogrisel you have an older pep8 ;) I needed to add it to my config file.

xuewei4d commented:
![demo_bgmm_full](https://cloud.githubusercontent.com/assets/1180956/9016583/2a5fbab6-37a0-11e5-9cf2-9549b6060cb7.gif)

![demo_bgmm_tied](https://cloud.githubusercontent.com/assets/1180956/9016589/327fdf1e-37a0-11e5-93e1-d852bb874f5d.gif)

![demo_bgmm_diag](https://cloud.githubusercontent.com/assets/1180956/9016591/36a0f9ac-37a0-11e5-86d4-67274ebc5aa3.gif)

![demo_bgmm_spherical](https://cloud.githubusercontent.com/assets/1180956/9016593/3a906944-37a0-11e5-82ed-cc7cd3bab78c.gif)

hi, I fixed some bugs in the computation of lower bound. Here it is new (probably bug-free) visualization of ```BayesianGaussainMixture```. 

```BayesianGaussianMixture``` with 'tied'  precision may diminish the the ability of automatically selecting suitable number of components, since the nearly starved components will contain the information of others  via shared precision/covariance. Therefore, the initial methods 'kmeans' which scatters components would be suitable.

amueller commented:
I don't understand the results from the spherical plot. Why are the pink and green components so much wider? Is that the prior?

amueller commented:
Can we do an example with an animation btw?

xuewei4d commented:
> I don't understand the results from the spherical plot. Why are the pink and green components so much wider? Is that the prior?

Yes. I think so. It may due to the pink and the green components don't have many data points at the last. So the prior information plays a major role in the estimation of precision/covariance. In Equation 4.31,  N_k is small, lambda_k is small and m_0 is the initial mean of components, so the last term in the equation is much larger. @amueller

xuewei4d commented:
> Can we do an example with an animation btw?

What kind of animation else would you like?

xuewei4d commented:
> "PEP8 said about these lines W503 line break before binary operator" I disagree with that. @ogrisel you have an older pep8 ;) I needed to add it to my config file.

Yeah. I saw the PEP8 warning, but I am afraid that putting the binary operators at the end of lines may make other people feel tied to check the codes agains with equations.   I don't know. @amueller

xuewei4d commented:
Hi, I cannot make BGMM with 'tied' and 'spherical' precision find correct mixtures on the very simple toy data ```RandData``` I created in test_gaussianmixture. As I said, with 'tied' precision, the shared 'tied' precision / covariance undermine the ability of selecting the suitable number of components. BGMM with 'spherical' may be too simple, it tries to fit a spherical distribution on the boundary between two distinct components. @ogrisel @amueller @lesteve

xuewei4d commented:
> Sorry for the prolonged silence. I'm just catching up with the issues and I hope to have a look at your PR soon.

Thanks! looking forward to any comment.

xuewei4d commented:
> make the tests pass and ensure at least 95% coverage of the code
> update the mixture documentation
> update the mixture examples
> benchmark the peformance of GaussianMixture vs the old GMM: we should not have any performance regression (time-wise and space-wise).
> For me this is the priority.

New BGMM has 97% coverage now. Could you let me know where should I write the mixture documentation and examples? I will test the performance now. @ogrisel

xuewei4d commented:
I did the profiling on ```gmm``` and ```Gaussianmixture```. Initialized with the same parameters, they use the same amount of time to converge.

amueller commented:
That's great.
The animations I thought might be good for the examples are animations like the ones you did above. I just wasn't sure if that is easy to do using plain matplotlib.

xuewei4d commented:
I see. Yeah, I think we can do this. The package I used to draw is matplotlib.animation.

xuewei4d commented:
Another question is the measure of the log likelihood / lower bound.

The log likelihood by definition should be sum of log probability of each sample under the current mixture   with i.i.d assumption. When the iteration should be stopped should depend on whether the log likelihood changes within the tolerance or not.  In ```BayesianGaussianMixture```, it is not just the sum of log probabilities of samples, but plus several terms involving parameters w.r.t variational distributions. However, ```GMM``` use ```mean``` instead  of ```sum``` of log probabilities, [link](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/mixture/gmm.py#L520). If we use ```mean``` on the data-dependent terms in the lower bound of BGMM, it would make the parameter-dependent terms play a main role, which I think it is not suitable. So I would prefer to go with ```sum``` of log probabilities in GMM, [link](https://github.com/scikit-learn/scikit-learn/pull/4802/files#diff-e9d6137cee72f77fd1b3f3ef16fc4dabR621)

xuewei4d commented:
I just got the function of printing verbose messages back. Now it has least correlation with ```_fit```. And it looks better than #4653. The extra parameter ```verbose_interval``` is to prevent printing too much Iteration messages.

When verbose = 1,
```

Initialization 0
  Iteration 0
  Iteration 10
  Iteration 20
  Iteration 30
  Iteration 40
Initialization converged: True
Initialization 1
  Iteration 0
  Iteration 10
  Iteration 20
  Iteration 30
  Iteration 40
Initialization converged: True
```

When verbose = 2, 
```
Initialization 0
  Iteration 0	 time lapse 0.00677s	 ll change inf
  Iteration 10	 time lapse 0.03628s	 ll change 36.30407
  Iteration 20	 time lapse 0.02995s	 ll change 27.55574
  Iteration 30	 time lapse 0.02990s	 ll change 20.46238
  Iteration 40	 time lapse 0.02776s	 ll change 0.00000
Initialization converged: True	 time lapse 0.13692s	 ll -329.36254
Initialization 1
  Iteration 0	 time lapse 0.00385s	 ll change inf
  Iteration 10	 time lapse 0.02949s	 ll change 2.88677
  Iteration 20	 time lapse 0.02683s	 ll change 62.57546
  Iteration 30	 time lapse 0.02957s	 ll change 33.31402
  Iteration 40	 time lapse 0.02520s	 ll change 0.00002
Initialization converged: True	 time lapse 0.12137s	 ll -329.36254
```

xuewei4d commented:
I am editing the ```http://scikit-learn.org/stable/modules/mixture.html```, but it seems the doc didn't mention BGM's ability of automatically selecting the number of components, and this ability coincides with DPGM's. I am wondering the advantage of DPGM using variational inference.

xuewei4d commented:
May I add my derivation in PDF to the doc?

ogrisel commented:
When I run the full test suite in this branch I get seemingly unrelated test failure. They disapear when I rebase the branch on the current master. Could you please rebase the branch (and collapse all intermediate commits at the same time)?

ogrisel commented:
> May I add my derivation in PDF to the doc?

Please convert it from latex to Restructured Text to include it in the doc (maybe as a new file referenced from the mixture chapter). You can use the latex math syntax to include formula in restructured text so there should not be much to edit.

ogrisel commented:
> "PEP8 said about these lines W503 line break before binary operator"

Fair enough, I don't care about this warning either, I updated my flake8 linter.

ogrisel commented:
> I don't understand the results from the spherical plot. Why are the pink and green components so much wider?

Indeed I don't find this behavior intuitive either.

xuewei4d commented:
> When I run the full test suite in this branch I get seemingly unrelated test failure. They disapear when I rebase the branch on the current master. Could you please rebase the branch (and collapse all intermediate commits at the same time)?

Do you mean collapse them into only one commit?

amueller commented:
The prior is different when using infinitely many components even when approximating it with finitely many.

ogrisel commented:
> Do you mean collapse them into only one commit?

yes. unless some commits are particularly interesting in isolation but I don't think it's the case.

ogrisel commented:
you could have one commit for all the code + tests, one for each example and one for the doc for instance.

superbobry commented on sklearn/mixture/base.py:
```diff
@@ -0,0 +1,454 @@
+                # print verbose msg
+                self._print_verbose_msg_iter_end(i, change)
+
+                if change < self.tol:
```
Convergence criterion along with the `_snapshot` functionality can be abstracted into a separate `ConvergenceMonitor` class, which will take care of both monitoring and reporting the fitting progress. This will hopefully make the `BaseMixture` interface cleaner.

xuewei4d replied:
If I use ```ConvergenceMonitor```, I still have to insert codes like 
``` Python
ConvergenceMonitor._print_verbose_msg_iter_end(i, change)
```
into ```fit``` right? @superbobry

superbobry replied:
Not necessarily. I'm experimenting with this approach in `hmmlearn`. Here's how the EM loop [looks](https://github.com/hmmlearn/hmmlearn/blob/master/hmmlearn/base.py#L418) in the current version:

```python
for iter in range(self.n_iter):
    curr_logprob = ...
    # ...
    self.monitor_.report(curr_logprob)
    if self.monitor_.converged:
        break
```

superbobry replied:
@xuewei4d have you considered implementing the convergence monitor as a separate entity?

superbobry commented on sklearn/mixture/base.py:
```diff
@@ -0,0 +1,454 @@
+
+                # e step
+                current_log_likelihood, resp = self._e_step(X)
+                change = abs(current_log_likelihood - prev_log_likelihood)
```
You might want to avoid the `abs`  here, because the change in LL should always be positive.

xuewei4d replied:
Thanks, I would let it raise ConvergenceWarning if change < 0. @superbobry @ogrisel @amueller @lesteve

xuewei4d replied:
I tested some cases with BayesianGaussianMixture. I think it might because of numerical issue when some mixtures starve (cannot cover enough data points) with 'tied' and precision, the change of LL would be negative. Other precision cases works fine. It might because 'tied' precision is not suitable for BayesianMixture

superbobry replied:
In my experience the negative change in LL is almost always a sign of a mistake somewhere in the code. Please double check the update for these covariance types.

xuewei4d commented:
> you could have one commit for all the code + tests, one for each example and one for the doc for instance.

OK. Sorry, but when I wrote the doc I found some bugs and committed along with the doc commit . :|

superbobry commented on sklearn/mixture/dpgaussianmixture.py:
```diff
@@ -0,0 +1,245 @@
+        'random', None. If it is not None, the responsibilities and model
+        parameters are initialized by the chosen method.
+
+    gamma_prior: float, default to 0.1.
```
Probably alpha in the description should be gamma. 

Also, the original [paper](http://www.cs.columbia.edu/~blei/papers/BleiJordan2004.pdf) by Blei and Jordan discussed a way of putting a prior on gamma (see Appendix B). Have you considered implementing it?

xuewei4d replied:
No, I didn't. But I am wondering why the expected number of mixtures of DPGM does not respond the different value of ```gamma_prior```.

superbobry replied:
Sorry if I've misunderstood your notation, but I think the expected number of mixture components grows as `alpha_prior * log(N)`.

xuewei4d replied:
Yes. I noticed the typo in the description.

xuewei4d commented:
I have uploaded an [IPN](https://github.com/xuewei4d/ipn/blob/master/scikit-learn/bgm_test.ipynb) about ```BayesianGaussianMixture```. With different random state, usually with 'kmeans' initialization, the lower bound would go down at the first few iterations. With 'random' initialization, the lower bound would oscillate down at the last few iterations. I double checked the code I wrote, but they seem OK.

xuewei4d commented:
Another problem is about ```DirichletProcessGaussianMixture```. Whatever I set ```gamma_prior```, the actual number of components after iteration does not change.

ogrisel commented:
> OK. Sorry, but when I wrote the doc I found some bugs and committed along with the doc commit . :|

No pbm, squash everything into one commit for now.

Note: it's always possible to use `git reset` to undo a commit and create new ones with the changes you want.

ogrisel commented on sklearn/mixture/bayesianmixture.py:
```diff
@@ -0,0 +1,985 @@
+
+
+def xlogx(x):
+    return np.sum(x[x < np.e]) + np.sum(x[x > np.e])
```
This is never used.

xuewei4d replied:
Sorry, this was used for testing the underflow problem. I accidentally committed.

ogrisel commented:
The coverage for the new files is a bit weak for the classical gaussian mixture and the base class:

```
sklearn.mixture.base                               171     21     53      6    88%   131, 136, 150, 166, 224, 263, 279-283, 343, 362, 367, 429, 433, 437, 441, 445, 449, 453, 457
sklearn.mixture.bayesianmixture                    381     11    100      8    96%   13, 246, 657-658, 686-687, 917, 975, 978-979, 984
sklearn.mixture.dpgaussianmixture                   59      1      4      1    97%   168
sklearn.mixture.gaussianmixture                    194     23     44     13    85%   330-331, 362-363, 389, 415, 574, 578, 582, 667-677, 690, 704, 708
```

The last columns list the line numbers that have no test coverage at all. You can ignore lines that do print only in verbose mode and the pass statements of abstract methods but I think there are some methods that should be tested.

You can re-run the coverage tests with `make test-coverage`. This will also generate a HTML version of the coverage report in `coverage/index.html` that is easier to browse that the text report.

ogrisel commented:
To be consistent with the rest of the library it would be better to rename the `n_iter` constructor param to `max_iter` and store the effective number of performed iterations (before convergence if reached) as a new `n_iter_` attribute on the estimators. `git grep max_iter` / `git grep n_iter_` to see other examples of this pattern in the code base.

ogrisel commented:
As we already do `n_init` for the mixture model, I think we should pass `n_init=1` to the underlying `KMeans` constructor to avoid performing `n_init x 10` calls to `KMeans.fit`.

xuewei4d committed:
minor improvement

amueller commented:
#5344 might be of interest

GaelVaroquaux commented on sklearn/mixture/__init__.py:
```diff
@@ -7,10 +7,18 @@
 
+from .gaussianmixture import GaussianMixture
+from .bayesianmixture import BayesianGaussianMixture
+from .dpgaussianmixture import DirichletProcessGaussianMixture
```
I would prefer the files to be named "gaussian_mixture", "bayesian_mixture" and "dp_gaussian_mixture".

GaelVaroquaux commented on sklearn/mixture/base.py:
```diff
@@ -0,0 +1,458 @@
+from .. import cluster
+from ..utils import check_random_state, check_array
+from ..utils.extmath import logsumexp
+from sklearn.utils import ConvergenceWarning
```
It would be better to have a relative import here.

xuewei4d committed:
rename file names, fix relative import, PEP8

xuewei4d commented:
Just renamed the file names, fixed relative import issue, and fixed PEP8 problems in test files

rvraghav93 commented:
Could you rebase? :)

superbobry commented:
I've looked through the current documentation (and derivation PDF) and I think it's missing the explanation of different covariance types.

We had a similar issue in `hmmlearn` recently, so feel free to use the explanation in hmmlearn/hmmlearn#60.

xuewei4d commented:
@superbobry Thanks!

xuewei4d committed:
DOC: add the comments on the types of covariances, fix typo

ogrisel commented:
@xuewei4d did you get the opportunity to understand what's wrong with the `gamma_prior` of your implementation of DP mixture models not working as expected (no impact on the number of components)?

In your notebook, you should seed the `random_state=0` parameter of the model to make different  runs easier to compare. Also the debug snapshot should call `.copy()` on all the arrays it snapshot. However this is not the cause of the problem.

tguillemot commented on sklearn/mixture/bayesian_mixture.py:
```diff
@@ -0,0 +1,980 @@
+            mk_sol[k] = np.sum(np.square(sol))
+        temp2 = self.mu_beta_prior_ * self.lambda_nu_ * mk_sol
+        temp_mu = (.5 * np.sum(temp1 - temp2) +
+                   self._log_gaussian_norm_prior)
```
My previous comment wasn't clear enough. I give more details here to be sure, we are ok.
I thought :
```
temp_mu = (.5 * np.sum(temp1 - temp2) +
                        self.n_components * self._log_gaussian_norm_prior)
```
According to Line532, `self._log_gaussian_norm_prior = D ln (beta_0 / (2pi))`
According to the formula 7.9 of your report, you do : `sum_{k=1^K}  (D ln ( beta/(2 pi))` 
Normaly on Line789, it should be : `self.n_components * self._log_gaussian_norm_prior`
Am I right ?

ogrisel replied:
@xuewei4d if you have some spare time could you please confirm this?

xuewei4d replied:
Oh, YES. @tguillemot . There should be ```self.n_components * ``` before ```self._log_gaussian_norm_prior```

tguillemot replied:
Ok it was sure I haven't miss something. Thx @xuewei4d .

xuewei4d commented:
I just come out the possible reason for the lower bound of VBGMM goes up and down. In VBGMM and DPGMM, some clusters may vanish during fitting, so the actual the number of clusters would not be the parameter ```self.n_components```. Maybe the value should be adjusted when some clusters are gone. I not quite sure. What I mean 'the clusters are gone' is that these clusters do not have enough data points to make its covariance positive definite. @tguillemot

superbobry commented:
@xuewei4d in my experience oscillating lower bound almost always means a bug in the M-step (or posterior update in case of VB-type algorithms).

> What I mean 'the clusters are gone' is that these clusters do not have enough data points to make its covariance positive definite.

Wouldn't the covariance be completely defined by the prior in this case?

amueller commented:
This might actually provide a good reference: https://github.com/teodor-moldovan/dpcluster

tguillemot commented:
@amueller Thanks I will have a look to that.

ogrisel commented:
Closing in favor of #7295. Let's finish the review there.
